### ISSUE TITLE
GCP PubSub Source does not validate certificates correctly

### ISSUE THREAD

**Issue by proffalken** [vectordotdev/vector#12590]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

When using the new GCP PubSub *source* (as introduced in #12057 ) in the nightly builds, Vector issues a transport error:

```
vector_1                | 2022-05-04T15:49:52.457046Z ERROR source{component_kind="source" component_id=source_gcp_pubsub component_type=gcp_pubsub component_name=source_gcp_pubsub}: vector::sources::gcp_pubsub: Source failed. error=Could not connect channel: transport error
```
enabling full debug provides the following logs:

```
vector_1                | 2022-05-04T15:49:52.421921Z DEBUG source{component_kind="source" component_id=source_gcp_pubsub component_type=gcp_pubsub component_name=source_gcp_pubsub}: hyper::client::connect::http: connecting to 142.250.179.234:443
vector_1                | 2022-05-04T15:49:52.435410Z DEBUG source{component_kind="source" component_id=source_gcp_pubsub component_type=gcp_pubsub component_name=source_gcp_pubsub}: hyper::client::connect::http: connected to 142.250.179.234:443
vector_1                | 2022-05-04T15:49:52.435490Z DEBUG source{component_kind="source" component_id=source_gcp_pubsub component_type=gcp_pubsub component_name=source_gcp_pubsub}: rustls::client::hs: No cached session for DnsName(DnsName(DnsName("pubsub.googleapis.com")))    
vector_1                | 2022-05-04T15:49:52.435556Z DEBUG source{component_kind="source" component_id=source_gcp_pubsub component_type=gcp_pubsub component_name=source_gcp_pubsub}: rustls::client::hs: Not resuming any session    
vector_1                | 2022-05-04T15:49:52.456617Z DEBUG source{component_kind="source" component_id=source_gcp_pubsub component_type=gcp_pubsub component_name=source_gcp_pubsub}: rustls::client::hs: Using ciphersuite Tls13(Tls13CipherSuite { suite: TLS13_AES_256_GCM_SHA384, bulk: Aes256Gcm })    
vector_1                | 2022-05-04T15:49:52.456657Z DEBUG source{component_kind="source" component_id=source_gcp_pubsub component_type=gcp_pubsub component_name=source_gcp_pubsub}: rustls::client::tls13: Not resuming    
vector_1                | 2022-05-04T15:49:52.456822Z DEBUG source{component_kind="source" component_id=source_gcp_pubsub component_type=gcp_pubsub component_name=source_gcp_pubsub}: rustls::client::tls13: TLS1.3 encrypted extensions: [Protocols([PayloadU8([104, 50])])]    
vector_1                | 2022-05-04T15:49:52.456846Z DEBUG source{component_kind="source" component_id=source_gcp_pubsub component_type=gcp_pubsub component_name=source_gcp_pubsub}: rustls::client::hs: ALPN protocol is Some(b"h2")    
vector_1                | 2022-05-04T15:49:52.456908Z  WARN source{component_kind="source" component_id=source_gcp_pubsub component_type=gcp_pubsub component_name=source_gcp_pubsub}: rustls::conn: Sending fatal alert BadCertificate    
vector_1                | 2022-05-04T15:49:52.457046Z ERROR source{component_kind="source" component_id=source_gcp_pubsub component_type=gcp_pubsub component_name=source_gcp_pubsub}: vector::sources::gcp_pubsub: Source failed. error=Could not connect channel: transport error
vector_1                | 2022-05-04T15:49:52.457064Z ERROR source{component_kind="source" component_id=source_gcp_pubsub component_type=gcp_pubsub component_name=source_gcp_pubsub}: vector::topology: An error occurred that Vector couldn't handle.
vector_1                | 2022-05-04T15:49:52.457082Z DEBUG source{component_kind="source" component_id=source_gcp_pubsub component_type=gcp_pubsub component_name=source_gcp_pubsub}: vector::topology::builder: Source pump finished.
```

Visiting https://pubsub.googleapis.com returns a valid certificate with the appropriate domain in the SAN.

Switching to a "regional" domain such as `us-central-2` returns an invalid certificate, as does vising the IP address, however I believe that this is a deliberate configuration by Google rather than a misconfiguration.

Setting `tls.verify_certificate` to `false` also does not make a difference.

### Configuration

```text
sources:
  source_gcp_pubsub:
    type: gcp_pubsub
    project: ${GCP_PROJECT}
    subscription: ${GCP_SUBSCRIPTION}
    endpoint: "https://us-central2-pubsub.googleapis.com"
    credentials_path: /google_cloud_service_account.json
    tls:
      verify_certificate: false
```

### Version

vector 0.22.0 (x86_64-unknown-linux-musl 46ab99e 2022-05-03)

### Debug Output

```text
No backtrace is created upon running this command
```

### Additional Context

Vector running in GCP Cloud Run, image is based on Nightly from 2022-05-03

### References

#12057

**Comment by fraserdarwent** [comment]
@spencergilbert I believe this is related to the gcp_pubsub source, not the sink as you've applied the label for

**Comment by spencergilbert** [comment]
Yep ☹️ - the source label hasn't been added yet

**Comment by spencergilbert** [comment]
👀 got the right label this time 😆

**Comment by proffalken** [comment]
This was fixed by #12594
